src/map/lua/lua_baseentity.cpp

```diff
@@ -3792,16 +3794,28 @@ inline int32 CLuaBaseEntity::addShopItem(lua_State *L)
     TPZ_DEBUG_BREAK_IF(m_PBaseEntity == nullptr);
     TPZ_DEBUG_BREAK_IF(m_PBaseEntity->objtype != TYPE_PC);
 
-    TPZ_DEBUG_BREAK_IF(lua_isnil(L, -1) || !lua_isnumber(L, -1));
-    TPZ_DEBUG_BREAK_IF(lua_isnil(L, -2) || !lua_isnumber(L, -2));
+    TPZ_DEBUG_BREAK_IF(lua_isnil(L, 1) || !lua_isnumber(L, 1));
+    TPZ_DEBUG_BREAK_IF(lua_isnil(L, 2) || !lua_isnumber(L, 2));
 
-    uint16 itemID = (uint16)lua_tonumber(L, -2);
-    uint32 price = (uint32)lua_tonumber(L, -1);
+    uint16 itemID = (uint16)lua_tonumber(L, 1);
+    uint32 price = (uint32)lua_tonumber(L, 2);
 
     uint8 slotID = ((CCharEntity*)m_PBaseEntity)->Container->getItemsCount();
-
     ((CCharEntity*)m_PBaseEntity)->Container->setItem(slotID, itemID, 0, price);
 
+    if(lua_isnumber(L, 3) && lua_isnumber(L, 4))
+    {
+        uint8 guildID = (uint8)lua_tonumber(L, 3);
+        uint16 guildRank = (uint16)lua_tonumber(L, 4);
+
+        // The rank numbers in status.lua are fake. Internally, rank is really
+        // a number between 0->1600. We fix the number here but really should solve
+        // the root problem "some day" but it will need a rank system rewrite.
```

Comment:
So, the internal rank is just the max skill level attainable at the current rank. The rank numbers in status.lua are just for our convenience and should probably be left as-is since they're used in a few different calculations so it would just shift around where these adjustments are, resulting in needing to do stuff like changing:
```lua
function getAdvImageSupportCost(player,craftID)
    local Rank = player:getSkillRank(craftID)
    return (Rank/100)*30
end
```
to:
```lua
function getAdvImageSupportCost(player,craftID)
    local Rank = player:getSkillRank(craftID)
    return (Rank+1)*30
end
```
So just rearranging where we're doing calculations instead of actually fixing/improving anything. I would just remove this comment. Everything else looks good I think.

Reply:
My intention is to point out why the number was being modified in such a counter-intuitive manner. Ideally, I prefer not having such artificial abstractions, as they cause this kind of confusion whenever that boundary needs to be crossed. It would also be a very simple manner to just redefined the status.lua values to something representing their *real* value, so I don't think the comment is invalid. But I will change it as you request.

Reply:
It's not counter intuitive though, if your rank is 1, your skill cap is 20. These calculations aren't converting rank into the retail accurate version of rank, they're calculating the max skill level available at your rank.

Reply:
In the image attached you can see the rank being passed by the packet is at 0, 200, 300, 400, 500 (0 is not joined a guild, 100 is the first rank, i think is amateur)

Hope this helps with the confusion on why its increments of 100.

![image](https://user-images.githubusercontent.com/26943220/87513536-b7efde80-c62d-11ea-89e7-d5c278444645.png)

thanks for the review coco :)

Reply:
It's not actually the rank being passed though, its the skill cap which is derived from the rank.

Reply:
Actually it might even make sense to move the
```cpp
        guildRank = (guildRank + 1) * 100;
```
from addShopItem and doing that calculation in the packet like so
```cpp
        ref<uint16>(i * 12 + 0x12) = (PChar->Container->getGuildRank(slotID) + 1) * 100;
```

Reply:
I disagree with pushing the reality of the underlying rank value so deep into the stack that it's virtually unknowable. I can't help but imaging that at some point, the obfuscating *16-value-rank metaphor* we adopt internally will lead to issues. That could be someone running into a problem they can't solve due to values not being what they expect, and so may believing (perhaps based on completely accurate packet captures, as Kain did) that they need to write new functions to workaround an imagined limitation.

Reply:
Rank just corresponds to Amateur, Recruit, Initiate, Novice, Apprentice, Journeyman, Craftsman, Artisan, Adept, Veteran, and Expert. The other ranks are not implemented, but that's what a rank is. The value being used in the shop packet is clearly the skill cap corresponding to the rank, why would it make sense to rank up by 100 ranks every time you rank up when a rank is those things listed? Likewise the price of certain things are determined by a different calculation using the rank, and the guild turn in item also determines the item of the day by indexing with the rank. It is very possible PVLV has the [wrong label](https://github.com/ZeromusXYZ/PacketViewerLogViewer/blob/00ffbcc66a373a811255290f0d0a221eed09fdc6/helpers/PacketParsers.cs#L1517), it wouldn't be the first time. There's a lot of stuff that's still unknown or just best guess.